lib/pure/ioselectors.nim

```diff
@@ -0,0 +1,1849 @@
+      data*: T
+
+    SelectorKey[T] = object
+      ident : int
+      flags : set[Event]
+      param : int
+      key : ReadyKey[T]
+
+  when not defined(windows):
+    when hasThreadSupport:
+      type
+        SharedArrayHolder[T] = object
+          part: array[16, T]
+        SharedArray {.unchecked.}[T] = array[0..100_000_000, T]
+
+      proc allocSharedArray[T](nsize: int): ptr SharedArray[T] =
+        let holder = cast[ptr SharedArrayHolder[T]](
+                       allocShared0(sizeof(T) * nsize)
+                     )
+        result = cast[ptr SharedArray[T]](addr(holder.part[0]))
+
+      proc deallocSharedArray[T](sa: ptr SharedArray[T]) =
+        deallocShared(cast[pointer](sa))
+
+    template setNonBlocking(fd) =
+      var x: int = fcntl(fd, F_GETFL, 0)
+      if x == -1: raiseOSError(osLastError())
+      else:
+        var mode = x or O_NONBLOCK
+        if fcntl(fd, F_SETFL, mode) == -1:
+          raiseOSError(osLastError())
+
+    template setKey(s, f1, f2, e, p, d) =
+      s.fds[f1].ident = f1
+      s.fds[f1].flags = e
+      s.fds[f1].param = p
+      s.fds[f1].key.fd = f2
+      s.fds[f1].key.data = d
+
+    template clearKey(s, f) =
+      s.fds[f].ident = 0
+      s.fds[f].flags = {}
+
+  when supportedPlatform:
+    template blockSignals(newmask: var Sigset, oldmask: var Sigset) =
+      when hasThreadSupport:
+        if posix.pthread_sigmask(SIG_BLOCK, newmask, oldmask) == -1:
+          raiseOSError(osLastError())
+      else:
+        if posix.sigprocmask(SIG_BLOCK, newmask, oldmask) == -1:
+          raiseOSError(osLastError())
+
+    template unblockSignals(newmask: var Sigset, oldmask: var Sigset) =
+      when hasThreadSupport:
+        if posix.pthread_sigmask(SIG_UNBLOCK, newmask, oldmask) == -1:
+          raiseOSError(osLastError())
+      else:
+        if posix.sigprocmask(SIG_UNBLOCK, newmask, oldmask) == -1:
+          raiseOSError(osLastError())
+  #
+  # BSD kqueue
+  #
+  # I have tried to adopt kqueue's EVFILT_USER filter for user-events, but it
+  # looks not very usable, because of 2 cases:
+  # 1) EVFILT_USER does not supported by OpenBSD and NetBSD
+  # 2) You can't have one event, which you can use with many kqueue handles.
+  # So decision was made in favor of the pipes
+  #
+  when bsdPlatform:
+    const
+      MAX_KQUEUE_CHANGE_EVENTS = 64
+      MAX_KQUEUE_RESULT_EVENTS = 64
+
+    when hasThreadSupport:
+      type
+        SelectorImpl[T] = object
+          kqFD : cint
+          maxFD : uint
+          changesTable: array[MAX_KQUEUE_CHANGE_EVENTS, KEvent]
+          changesCount: int
+          fds: ptr SharedArray[SelectorKey[T]]
+          count: int
+          changesLock: Lock
+    else:
+      type
+        SelectorImpl[T] = object
+          kqFD : cint
+          maxFD : uint
+          changesTable: array[MAX_KQUEUE_CHANGE_EVENTS, KEvent]
+          changesCount: int
+          fds: seq[SelectorKey[T]]
+          count: int
+
+    when hasThreadSupport:
```

Comment:
Is that wise though? Should the exported API change depending on Nim's enabled thread support?